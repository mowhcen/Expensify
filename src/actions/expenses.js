import database from "../firebase/firebase";

// ADD_EXPENSE
const addExpense = (expense) => ({
    type: "ADD_EXPENSE",
    expense,
});

const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = "",
            note = "",
            amount = 0,
            createAt = 0,
        } = expenseData;

        const expense = { description, note, amount, createAt };

        return database
            .ref("expenses")
            .push(expense)
            .then((ref) => {
                dispatch(
                    addExpense({
                        id: ref.key,
                        ...expense,
                    })
                );
            });
    };
};

// REMOVE_EXPENSE
const removeExpense = ({ id } = {}) => ({
    type: "REMOVE_EXPENSE",
    id,
});
// EDIT_EXPENSE
const editExpense = (id, updates) => ({
    type: "EDIT_EXPENSE",
    id,
    updates,
});

// SET_EXPENSES
const setExpenses = (expenses) => ({
    type: "SET_EXPENSES",
    expenses,
});

const startSetExpenses = () => {
    return (dispatch) => {
        return database
            .ref("expenses")
            .once("value")
            .then((snapshot) => {
                const expenses = [];

                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val(),
                    });
                });

                dispatch(setExpenses(expenses));
            });
    };
};

export {
    removeExpense,
    editExpense,
    addExpense,
    startAddExpense,
    setExpenses,
    startSetExpenses,
};
