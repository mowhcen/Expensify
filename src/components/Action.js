import React from "react";

const Action = (props) => {
    return (
        <div>
            <button onClick={props.handlePick} disabled={!props.hasOption}>
                What Should I do?
            </button>
        </div>
    );
};

export default Action;
