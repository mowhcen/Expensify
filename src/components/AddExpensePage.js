import React from "react";

const AddExpensePage = () => <div>this from my add expense component</div>;

export default AddExpensePage;
