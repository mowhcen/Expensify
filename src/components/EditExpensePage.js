import React from "react";

const EditExpensePage = () => {
    return (
        <div>
            this from my edit Expense component number {props.match.params.id}
        </div>
    );
};
export default EditExpensePage;
