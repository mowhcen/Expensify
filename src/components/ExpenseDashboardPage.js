import React from "react";

const ExpenseDashboardPage = () => <div>this from my dashboard component</div>;

export default ExpenseDashboardPage;
