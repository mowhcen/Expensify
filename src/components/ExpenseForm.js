import React from "react";
import { SingleDatePicker } from "react-dates";
import moment from "moment";

class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : "",
            note: props.expense ? props.expense.note : "",
            amount: props.expense
                ? (props.expense.amount / 100).toString()
                : "",
            createAt: props.expense ? moment(props.expense.createAt) : moment(),
            calendarFocused: false,
            error: "",
        };
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;

        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;

        this.setState(() => ({ note }));
    };
    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createAt) => {
        if (createAt) {
            this.setState(() => ({ createAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({
            calendarFocused: focused,
        }));
    };
    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.description || !this.state.amount) {
            this.setState(() => ({
                error: "Please fill the description and the amount",
            }));
        } else {
            this.setState(() => ({ error: "" }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createAt: this.state.createAt.valueOf(),
                note: this.state.note,
            });
        }
    };
    render() {
        return (
            <form className="form" onSubmit={this.onSubmit}>
                {this.state.error && (
                    <p className="form__error">{this.state.error}</p>
                )}
                <input
                    className="text-input"
                    type="text"
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                />
                <input
                    className="text-input"
                    type="text"
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                />
                <SingleDatePicker
                    date={this.state.createAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
                <textarea
                    className="text-area"
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                ></textarea>
                <div>
                    <button className="button">Save Expense</button>
                </div>
            </form>
        );
    }
}

export default ExpenseForm;
