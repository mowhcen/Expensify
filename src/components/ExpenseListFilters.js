import {
    setEndDate,
    setStartDate,
    setTextFilter,
    sortByAmount,
    sortByDate,
} from "../actions/filters";

import { DateRangePicker } from "react-dates";
import React from "react";
import { connect } from "react-redux";

class ExpenseListFilters extends React.Component {
    state = {
        calenderFocus: null,
    };
    onDatesChange = ({ startDate, endDate }) => {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };
    onFocusChange = (calenderFocus) => {
        this.setState(() => ({ calenderFocus }));
    };
    render() {
        return (
            <div>
                <input
                    type="text"
                    value={this.props.filters.text}
                    onChange={(e) => {
                        props.dispatch(setTextFilter(e.target.value));
                    }}
                />
                <select
                    value={this.props.filters.sortBy}
                    onChange={(e) => {
                        if (e.target.value === "date") {
                            this.props.dispatch(sortByDate());
                        } else if (e.target.value === "amount") {
                            this.props.dispatch(sortByAmount());
                        }
                    }}
                >
                    <option value="amount">Amount</option>
                    <option value="date">Date</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calenderFocus}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        filters: state.filters,
    };
};

export default connect(mapStateToProps)(ExpenseListFilters);
