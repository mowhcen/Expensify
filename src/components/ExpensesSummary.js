import { Link } from "react-router-dom";
import React from "react";
import { connect } from "react-redux";
import getExpenseTotal from "../selectors/expenses-total";
import getVisibleExpenses from "../selectors/expenses";
import numeral from "numeral";

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
    const expenseWord = expenseCount === 1 ? "" : "s";
    const formattedExpensesTotal = numeral(expensesTotal / 100).format(
        "$0,0.00"
    );
    return (
        <div className="page-header">
            <div className="content-container">
                <h1 className="page-header__title">
                    Viewing <span>{expenseCount}</span> expense
                    {expenseWord} totalling{" "}
                    <span>{formattedExpensesTotal}</span>
                </h1>
                <div className="page-header__actions">
                    <Link className="button" to="/create">
                        Add Expenses
                    </Link>
                </div>
            </div>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);
    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: getExpenseTotal(visibleExpenses),
    };
};

export default connect(mapStateToProps)(ExpensesSummary);
