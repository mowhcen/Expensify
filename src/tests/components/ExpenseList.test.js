import { ExpenseList } from "../../components/ExpenseList";
import React from "react";
import expenses from "../fixtures/expenses";
import { shallow } from "enzyme";

test("should render ExpenseList with expenses", () => {
    const wrapper = shallow(<ExpenseList expenses={expenses} />);
    expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseList with empty message", () => {
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});
