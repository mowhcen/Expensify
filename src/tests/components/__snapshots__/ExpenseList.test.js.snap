// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div>
  <p>
    No expenses
  </p>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={1234}
    createAt={0}
    description="Gum"
    id="1"
    key="1"
    note="sample note 1"
  />
  <ExpenseListItem
    amount={9876}
    createAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note="sample note 2"
  />
  <ExpenseListItem
    amount={4567}
    createAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note="sample note 3"
  />
</div>
`;
